src/metabase/lib/equality.cljc

```diff
@@ -104,106 +109,365 @@
+(defn- match-needle-to-haystack
+  ([squinty-needle squinty-haystack] (match-needle-to-haystack squinty-needle squinty-haystack 0))
+  ([squinty-needle squinty-haystack depth]
+   (when (seq squinty-needle)
+     (let [needle  (first squinty-needle)
+           matches (keep-indexed (fn [i squinty-hay]
+                                   (when (= needle (first squinty-hay))
+                                     i))
+                                 squinty-haystack)]
+       (if (empty? matches)
+         (recur (rest squinty-needle) (map rest squinty-haystack) (inc depth))
+         (do
+           (when (> (count matches) 1)
+             (log/warn (i18n/tru "Ambiguous match for {0}: got {1}" (pr-str needle) (pr-str matches))))
+           [(first matches) depth]))))))
+
+(defn- lowest-depth
+  "Given a list of [haystack-index [needle-index depth]] pairs, find the one with the lowest depth.
+  If there's more than one pair with that depth, throw an error.
+
+  Returns the `needle-index` with the lowest depth."
+  [pairs]
+  (let [pairs     (map second pairs) ; Drop the unnecessary [haystack-index ...] outer layer.
+        min-depth (reduce (fn [m [_needle-index depth]] (min m depth)) 100 pairs)
+        at-depth  (filter #(= (second %) min-depth) pairs)]
+    (if (= (count at-depth) 1)
+      (ffirst at-depth)
+      ;; TODO: This should be thrown as an exception rather than a warning, but currently there are some ambiguous
+      ;; columns returned in certain cases of nested queries. Eg. Orders joined to Products, then nest that query.
+      ;; Then for each column in Products, `visible-columns` returns two: one with `:source/card`, and one with
+      ;; `:source/implicitly-joinable`. For regular queries we use the IDs to de-dupe them, but the columns from the
+      ;; nested query have no `:id` or `:table-id`.
+      (do
+        (log/warn  (i18n/tru "Ambiguous match: needles at {0} matched a single haystack value"
+                             (pr-str (map first at-depth))))
+        ;; Arbitrarily returning the earlier of the two matches.
+        (ffirst at-depth)))))
 
-  This first looks for a matching ref with a strict comparison, then in increasingly less-strict comparisons until it
-  finds something that matches. This is mostly to work around bugs like #31482 where MLv1 generated queries with
-  `:field` refs that did not include join aliases even tho the Fields came from joined Tables... we still know the
-  Fields are the same if they have the same IDs.
+(defn find-closest-matches-for-refs
+  "For each of `needles` (which must be a ref) find the column or ref in `haystack` that it most closely matches. This
+  is meant to power things like [[metabase.lib.breakout/breakoutable-columns]] which are supposed to include
+  `:breakout-position` for columns that are already present as a breakout; sometimes the column in the breakout does not
```

Comment:
Isn't this a bad example? Is this function recommended when matching against columns?